Benchmark some table operations

ddn0 opened the pull request:
```
Results from my machine (Release, USE_ARCH=native, clang 9.0.1, Ubuntu
20.04):

  -----------------------------------------------------------------
  Benchmark                       Time             CPU   Iterations
  -----------------------------------------------------------------
  FilterNaive/1024             4198 ns         4198 ns       170578
  FilterNaive/262144         643727 ns       643729 ns         1073
  FilterNaive/67108864    277311578 ns    277308825 ns            2
  FilterVanilla/1024            709 ns          709 ns       971023
  FilterVanilla/262144       166596 ns       166593 ns         4155
  FilterVanilla/67108864   81804938 ns     81803578 ns            8
  FilterArrow/1024             5281 ns         5281 ns       137600
  FilterArrow/262144         161539 ns       161537 ns         4329
  FilterArrow/67108864     62700532 ns     62699211 ns           11
  FilterOp/1024               64215 ns        64215 ns        11338
  FilterOp/262144           9639486 ns      9639496 ns           77
  FilterOp/67108864      2384855406 ns   2384840099 ns            1

Naive: Arrow operations, materializing a new table for each expression
  operator

Vanilla: Extract array from Arrow table, perform operations in normal
  C++

Arrow: Use arrow::compute

FilterOp: Dummy version of using FilterOp to apply an expression to a
  table. Does not implement proper evaluation. It just evaluates an
  expression for each row.
```

bowu commented:
Interesting! It seems variable binding causes a big performance problem.

l-hoang commented on libquery/test/table-bench.cpp:
```diff
@@ -0,0 +1,202 @@
+  auto int_input_array = std::dynamic_pointer_cast<arrow::UInt32Array>(input);
+  GALOIS_LOG_ASSERT(int_input_array != nullptr);
+
+  for (auto _ : state) {
```
a comment explaining at a high level what this benchmark is doing (keeping rows with a value greater than half the number of rows) somewhere would help readers more quickly understand what they're reading

ddn0 replied:
added something to the top of the file. Let me know if this works for you.

l-hoang replied:
excellent, thanks for doing this

ddn0 commented:
> at the moment, this doesn't really tell me anything unless i'm missing the point of what you were testing

Think of `FilterVanilla` as the natural lower bound on the performance of table operations. It would be nice to be within spitting distance of that performance.

`FilterArrow` actually does do that, which suggests that this is not a crazy goal. And it uses the arrow interfaces....

>  filter op, as you mentioned, is doing something completely different. if you wanted to replicate the other kernels you would have to create a column of ascending numbers, then change the WHERE to something like this

I can add this if you want. I didn't actually know how to use the expression evaluator, so that's why I started with a dummy expression. In any event, `FilterOp` is still meaningful because it is measuring the cost of evaluating some predicate on a table.

>  filterop is also doing a bit more extra as it's constructing the op graph every time; not sure how much overhead that would add, though if it adds that much overhead as shown there then the design might need to be revisited

This is a 50X difference. For the larger input sizes, the benchmark only runs once.

l-hoang commented:
> This is a 50X difference. For the larger input sizes, the benchmark only runs once.

yeah I can't argue with that, that's pretty bad

i'm curious as to where that extra time is coming from though

ddn0 commented:
```
$ perf record libquery/test/table-bench --benchmark_filter=FilterOp/67
$ perf report
# ........  ...............  ......................  ........................................................................................................
.............................................................................................................................................................
.............................................................................................................................................................
.............................................................................................................................................................
.............................................................................................................................................................
.............................................................................................................................................................
.......................................................................................................................
#
    20.66%  table-bench      libgalois_query.so      [.] galois::FilterOp::Compute
    18.71%  table-bench      libgalois_query.so      [.] galois::ArithmeticTree::Evaluate
    14.37%  table-bench      libgalois_query.so      [.] galois::ExpressionClause::Evaluate
    13.33%  table-bench      libgalois_query.so      [.] galois::Table::Get
     5.38%  table-bench      libgalois_query.so      [.] std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std
::pair<std::__cxx11::basic_string<char, std::char_traits<
char>, std::allocator<char> > const, std::variant<bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int, int, unsigned long, long, float> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::variant<bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int, int, unsigned long, long, float> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::~_Hashtable
[...]
```

The power of using existing performance tools....

l-hoang commented:
oof, evaluate hurts it looks like

ddn0 commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -100,7 +100,6 @@ std::unique_ptr<ArithmeticTree> ParseClauseData(CypherASTNode ast_node) {
   // literal
   if (ast_node_type == CYPHER_AST_INTEGER) {
     int int_data = std::stoi(cypher_ast_integer_get_valuestr(ast_node));
-    GALOIS_LOG_VERBOSE("Parsed {} as the int", int_data);
```
deleted a bunch more prints. Let me know if I went too far.

ddn0 commented on libgalois/src/Statistics.cpp:
```diff
@@ -83,6 +83,13 @@ void StatManager::print(void) {
 void StatManager::printStats(std::ostream& out) {
   mergeStats();
+
+  if (intStats.cbegin() == intStats.cend() &&
```
One day, I'm going to fix stats for good.

l-hoang replied:
what else is still wrong with it 🤔

roshandathathri commented on libquery/test/table-bench.cpp:
```diff
@@ -0,0 +1,222 @@
+        std::dynamic_pointer_cast<arrow::UInt32Array>(incremented_array);
+    for (int64_t i = 0, n = int_incremented_array->length(); i < n; ++i) {
+      if (int_incremented_array->Value(i) > num_rows / 2) {
+        if (auto r = filtered_builder.Append(int_incremented_array->Value(i));
```
Just curious. Why not create a vector, push to it, and then build arrow array like you did for increment?

ddn0 replied:
I was curious on testing the worst possible interface to arrow, so I tried to keep this variant to element-wise operations.

roshandathathri replied:
Ok, makes sense.

roshandathathri commented on libquery/test/table-bench.cpp:
```diff
@@ -0,0 +1,222 @@
+  return array;
+}
+
+void FilterArrow(benchmark::State& state) {
```
Where is State defined?

ddn0 replied:
in `<benchmark/benchmark.h>` probably. This is from the 3rd party library: https://github.com/google/benchmark

roshandathathri commented on libquery/test/table-bench.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+// Unlike the other benchmarks, this one doesn't implement the same kernel
+// because I'm not sure how to specify it. Instead, just run a dummy load.
```
It would be good if @bowu or @l-hoang can change this function to do similar computation as the ones above.

ddn0 commented:
Updated based on all the feedback so far.

@bowu note I found and fixed a bug in FilterOp ("Fix typo in FilterOp")

@roshandathathri using the precise expression rather than a dummy filter has the lovely effect of decreasing the performance of FilterOp by another order of magnitude.

l-hoang commented:
> @roshandathathri using the precise expression rather than a dummy filter has the lovely effect of decreasing the performance of FilterOp by another order of magnitude.

disappointed, but not surprised

witchel commented:
I also have this on my dev branch.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -431,6 +431,14 @@ class KATANA_EXPORT PropertyGraph {
+    return EntityTypeManager(node_entity_type_manager_);
+  }
+
+  EntityTypeManager GetCopyOfEdgeTypeManager() const {
```
Instead, why don't we return a const-reference to the member variable and let the user do a copy of it?

scober replied:
That works for me. I got some tunnel vision about my specific use case.

roshandathathri reviewed (approved):
LGTM
